mbed_targets/mbed_targets.py

```diff
@@ -124,22 +124,27 @@ class UnsupportedMode(ToolsError):
     """The Database Mode is unsupported."""
 
 
-class MbedTargets(Collection):
-    """Base Interface to the Target Database."""
+class MbedTargets(Set):
+    """Base Interface to the Target Database.
 
-    def __init__(self, target_data: Iterable[dict]) -> None:
+    Don't instantiate an MbedTargets directly, this class is intended for subclassing.
+
+    Subclasses are expected to override __init__ to create an iterable of MbedTarget objects and pass it to
+    the parent initialiser via super().__init__.
+    """
+
+    @classmethod
+    def _from_iterable(cls: object, it: Iterable) -> "MbedTargets":
+        """Override Set._from_iterable to always return an MbedTargets instance and not a subclass instance."""
+        return MbedTargets(it)
```

Comment:
So this is needed because the constructors in the subclasses don't take an iterable. I guess the gotcha here is that if you add additional methods/attributes to the subclass the result of say a union action wouldn't contain instances of the subclass but of MbedTargets which wouldn't have these methods. 
Might be worth making this super clear so people coming later on don't do this accidentally.